### PR TITLE
Support NumPy 2

### PR DESCRIPTION
**Pull request by DanielYang59**
### Summary

Support NumPy 2, previously #184

- [ ] Loose numpy > 2 pin once test finished

### Test tracker

|         | Windows  | Ubuntu | MacOS |
|---------|----------|--------|-------|
| NumPy 1 |          |        |       |
| NumPy 2 |          |        |       |

---

Should we worry about [the following warning](https://github.com/CederGroupHub/chgnet/actions/runs/10832673983/job/30057609706?pr=202):

```
tests\test_encoders.py .................
!!!!!! the two directed edges are equal but this operation is not supposed to happen
!!!!!! the two directed edges are equal but this operation is not supposed to happen
tests\test_graph.py ........
```

### CODE REVIEW

**Comment by janosh** [comment]
incidentally, one way to get around the need to manually approve CI runs on every commit would be to merge a small PR instantly. if you want you could fix a few of the `DOC` errors

```
ruff check . --select DOC --preview
```

and submit a PR for that which we merge right away

File: pyproject.toml [hunk 1]
```diff
@@ -9,10 +9,11 @@ license = { text = "Modified BSD" }
 dependencies = [
     "ase>=3.23.0",
     "cython>=3",
-    "numpy>=1.26,<2",
+    # "numpy>=1.26",
+    "numpy>=2.0.0",
     "nvidia-ml-py3>=7.352.0",
-    "pymatgen>=2023.10.11",
-    "torch>=1.11.0",
```
[the remark this answers is not in the document]
Comment (janosh) [reply pyproject.toml:15]: i'm fine with dropping torch v1 support _if_ that actually helps us. sounds like it doesn't?

[the remark this answers is not in the document]
Comment (DanielYang59) [reply pyproject.toml:15]: I think using `torch >= 2.4.1` resolves the problem, things should be clearer to see now

**Comment by DanielYang59** [comment]
Looks like we're almost there! `graph` and `model` tests are all passing, looks like you have fixed `magmom` several times from the commit history?

```
================================== FAILURES ===================================
______________________ test_parse_vasp_dir_with_magmoms _______________________

tmp_path = WindowsPath('C:/Users/runneradmin/AppData/Local/Temp/pytest-of-runneradmin/pytest-0/test_parse_vasp_dir_with_magmo0')

    def test_parse_vasp_dir_with_magmoms(tmp_path: Path):
        with ZipFile(f"{ROOT}/tests/files/parse-vasp-with-magmoms.zip") as zip_ref:
            zip_ref.extractall(tmp_path)
        vasp_path = os.path.join(tmp_path, "parse-vasp-with-magmoms")
        dataset_dict = parse_vasp_dir(vasp_path, save_path=f"{tmp_path}/tmp.json")
    
        assert isinstance(dataset_dict, dict)
        assert len(dataset_dict["structure"]) > 0
        assert len(dataset_dict["uncorrected_total_energy"]) > 0
        assert len(dataset_dict["energy_per_atom"]) > 0
        assert len(dataset_dict["force"]) > 0
>       assert len(dataset_dict["magmom"]) == len(dataset_dict["force"])
E       assert 0 == 1
E        +  where 0 = len([])
E        +  and   1 = len([[[0.22258223, 0.16966156, -0.80105683], [0.22261809, 0.16943309, -0.80118544], [0.22275722, 0.16423232, -0.80168788], [0.22281114, 0.1641892, -0.80192072], [0.01342199, 0.01066812, -0.04840373], [0.01351635, 0.01079529, -0.04858565], ...]])

tests\test_vasp_utils.py:29: AssertionError
_________________________ test_dataset_from_vasp_dir __________________________

tmp_path = WindowsPath('C:/Users/runneradmin/AppData/Local/Temp/pytest-of-runneradmin/pytest-0/test_dataset_from_vasp_dir0')

    def test_dataset_from_vasp_dir(tmp_path: Path):
        with ZipFile(f"{ROOT}/tests/files/parse-vasp-with-magmoms.zip") as zip_ref:
            zip_ref.extractall(tmp_path)
        vasp_path = os.path.join(tmp_path, "parse-vasp-with-magmoms")
        dataset = StructureData.from_vasp(vasp_path, save_path=f"{tmp_path}/tmp.json")
        assert len(dataset.structures) > 0
        assert isinstance(dataset.structures[0], Structure)
        assert len(dataset.structures) == len(dataset.energies)
        assert len(dataset.structures) == len(dataset.forces)
        assert len(dataset.structures) == len(dataset.stresses)
>       assert len(dataset.structures) == len(dataset.magmoms)
E       TypeError: object of type 'NoneType' has no len()

tests\test_vasp_utils.py:83: TypeError

**Comment by janosh** [comment]
why are only the windows tests failing? is the value of `dataset.magmoms` platform dependent?

**Comment by DanielYang59** [comment]
You really had me here :) I haven't looked into the source code yet (thought that's a known issue), would have a look tomorrow.

That's a bit weird because looks like `magmom` is float, I was not expecting too many issues with float.

**Comment by DanielYang59** [comment]
Again: https://github.com/CederGroupHub/chgnet/blob/9281cf49cc844f03346f0a645ac68cf5ecab06ff/chgnet/utils/vasp_utils.py#L62-L64

**Comment by janosh** [comment]
> reverse_readfile

your current arch nemesis strikes again 🤣

**Comment by DanielYang59** [comment]
That's sad, looks like we cannot merge this until reverse readline is fixed, because this wouldn't be a working version for windows users. 

> your current arch nemesis strikes again 🤣

OK! my time off is over, going back to work tomorrow :)

**Comment by DanielYang59** [comment]
Any comment on [the following message](https://github.com/CederGroupHub/chgnet/actions/runs/10833715838/job/30061170790?pr=202) (seemingly something terrible)?

```
tests\test_encoders.py .................
!!!!!! the two directed edges are equal but this operation is not supposed to happen
!!!!!! the two directed edges are equal but this operation is not supposed to happen
tests\test_graph.py ........
```

**Comment by janosh** [comment]
> Any comment on [the following message](https://github.com/CederGroupHub/chgnet/actions/runs/10833715838/job/30061170790?pr=202) (seemingly something terrible)?

i think those messages are expected (maybe from a test meant to check that error message)

**Comment by janosh** [comment]
i'm fine with merging this and having a `main` branch where the `parse_vasp_dir` has issues on windows. we can wait with a package release until those are resolved

**Comment by BowenD-UCB** [comment]
Thanks a lot for the migration to np 2! @DanielYang59 @janosh 
I just fixed the message `test_graph.py` and will merge this PR now

File: pyproject.toml [hunk 2]
```diff
@@ -9,10 +9,12 @@ license = { text = "Modified BSD" }
 dependencies = [
     "ase>=3.23.0",
     "cython>=3",
-    "numpy>=1.26,<2",
+    # "monty==2024.7.12",  # TODO: restore once readline fixed
+    # "numpy>=1.26",  # TODO: remove after test
+    "numpy>=2.0.0",
```
Comment (DanielYang59) [inline pyproject.toml:14]: This pin should be removed, it's for test purpose, we're supporting NP2 not NP2-only 🤣

**Pull request merged**